Use zip32 if possible

rullzer opened the pull request:
fixes #7782 

* OSX doesn't handle 64zip that well
* Some other implentations don't handle it perfectly either
* If the file is belog 4GiB (some overhead) => zip32
* This covers the 99% case I bet


In some future work we should really move this to some decent Node based API. All this static stuff is making me cry 😭 

Requires:
- [x] https://github.com/nextcloud/server/issues/9101

rullzer committed:
Use zip32 if possible

* OSX doesn't handle 64zip that well
* Some other implentations don't handle it perfectly either
* If the file is belog 4GiB (some overhead) => zip32
* This covers the 99% case I bet

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

icewind1991 commented on lib/private/legacy/files.php:
```diff
@@ -193,6 +204,25 @@ public static function get($dir, $files, $params = null) {
+		$fileSize = 0;
+
+		if (\OC\Files\Filesystem::is_file($path)) {
+			$fileSize = \OC\Files\Filesystem::filesize($path);
```
please use `getFileInfo` instead

rullzer replied:
will do

icewind1991 commented on lib/private/legacy/files.php:
```diff
@@ -193,6 +204,25 @@ public static function get($dir, $files, $params = null) {
+		if (\OC\Files\Filesystem::is_file($path)) {
+			$fileSize = \OC\Files\Filesystem::filesize($path);
+		} elseif (\OC\Files\Filesystem::is_dir($path)) {
+			$files= \OC\Files\Filesystem::getDirectoryContent($path);
```
with `getFileInfo` you can just get the directory size in one go without the need to recurse

rullzer replied:
oeee

icewind1991 commented on lib/private/legacy/files.php:
```diff
@@ -142,17 +142,28 @@ public static function get($dir, $files, $params = null) {
+			if ($getType === self::ZIP_FILES) {
+				$fileSize = 0;
+				foreach ($files as $file) {
+					$fileSize += \OC\Files\Filesystem::getFileInfo($dir . '/' . $file)['size'];
```
`->size()` instead of `['size']` for bonus points

rullzer replied:
of course I want bonus points

MorrisJobke commented:
I tried it on macOS again and it still is not possible to extract the zip with the built-in tools. I also checked that the first if clause is used and this is the case.

rullzer commented:
:S stupid OSX. No clue then

danxuliu reviewed (commented):
This makes possible to open downloaded zip files with KDE Archive KIOSlave (that is, directly in a file explorer or in an _Open file_ dialog) :-D

I have taken a look to the [Zip specification](https://pkware.cachefly.net/webdocs/APPNOTE/APPNOTE-6.3.4.TXT) and it seems that it is safe to use zip32 for a basic (without compression, volumes nor encryption) zip file when all the following conditions are met (do not quote me on this, though ;-) ):
- No file size is larger than 4 bytes (file size < 4294967296); see _4.4.9 uncompressed size_
- The size of all files plus their local headers is not larger than 4 bytes; see _4.4.16 relative offset of local header_ and _4.4.24 offset of start of central directory with respect to the starting disk number_
- The total number of entries (files and directories) in the zip file is not larger than 2 bytes (number of entries < 65536); see _4.4.22 total number of entries in the central dir_
- The size of the central directory (similar to the local headers for each file) is not larger than 4 bytes; see _4.4.23 size of the central directory_

The first two are already covered in this pull request. The fourth one is mostly covered by the margin between 4x1000^3 and 4x1024^3; if I am not mistaken it is still technically possible to create a zip file from files smaller than 4GB with a central directory larger than 4GiB (which would be broken with zip32), but it should not happen in the real world.

However, it could be good to take into account the third case too (that is, use zip32 only if the size is less than 4GB and the total number of files and directories is less than 65536); it is also an unlikely scenario, yet I guess that this one could enter in the _plausible_ category.

rullzer commented:
@danxuliu thanks for the detailed comment :smile: 

I agree we should take this all properly into account.
I propose to get this in and abstract further away in https://github.com/nextcloud/server/pull/8017

That way we can just keep track of all the files/folders/file handles. And do all kind of calculations. Including the number of files and the size.

danxuliu committed:
Use zip32 only if there are less than 65536 files

A zip32 file can contain, at most, 65535 files (and folders), so take
that constraint into account.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add integration tests for downloading basic zip files

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add integration tests for zip32/zip64 boundaries of number of files

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu commented:
I have noticed a difference in behaviour from the old code to the new one: if one of the files to be downloaded does not exist, before the zip was downloaded anyway but without that file; now an error is shown to the user. I guess the new behaviour is fine, but I mention it just in case ;-)

>  I propose to get this in and abstract further away in #8017

I could not wait ;-) so I have addressed my own comment about using zip32 only if the number of files and directories is less than 65536

I have also added integration tests for downloading a zip file. The problem is that the tests for zip32/zip64 boundaries (65535/65536 files) are terribly slow; in my system they take more than two and a half hours... But it is a 8~9-years old machine, so let's see if Drone workers can finish them in a reasonable amount of time; if not they can just be disabled in Drone but kept to be manually run when working on zip related things.

Pending issues (for this pull request or future ones ;-) ):
- Until [the PHPZipStreamer fix for OSX is added](https://github.com/McNetic/PHPZipStreamer/pull/40) the `if` branches should be reordered to prefer Tar; if this pull request is merged once the OSX fix is added no changes are needed (should the tar branch be removed in that case? Or should it be kept to use Tar instead of zip64 in OSX?)
- Once the PHPZipStreamer fix for OSX is added the checks in the integration tests could be extended to look for the data descriptor header too
- Currently if the server runs on a 32 bits system and a zip64 should be generated (because the zip32 constraints are not met) [a zip32 file is generated anyway](https://github.com/nextcloud/server/pull/5116#issuecomment-307313470). That file will be broken, so an error telling something like "_This Nextcloud server can not generate zip64 files. Please contact your system administrator (or try to download less files)._" or something like that.
- Is the check against `4 * 1000 * 1000 * 1000` safe in 32 bit machines? [PHP integers overflow to floats](http://php.net/manual/en/language.types.integer.php#language.types.integer.overflow) so it is probably fine, but I am not sure. In any case, even if using zip32, can 32 bit machines generate 4GB zip files or are they limited to 2^31 bytes (as there are not unsigned integers in PHP)?
- The integration tests for zip64 should not run in 32 bit systems. Unfortunately Behat does not support skipping a test by throwing SkippedException or something like that; [the support for that needs to be explicitly added](https://tracker.moodle.org/browse/MDL-43330). A cheesy alternative would be to throw `PendingException` to mark the test step as pending and skip the rest of tests in the scenario... :-P In any case, once tests can be skipped it would be possible to add other integration tests that create a 4GB file, but checking beforehand if there is enough space in the file system to do that

MorrisJobke commented:
@danxuliu They fail - maybe rebase to get the latest .drone.yml update as well ;)

danxuliu commented:
@MorrisJobke The integration test fails because I naively used `../../data` to access the data directory... but of course that does not work if the data directory is elsewhere, like happens when run in Drone ;-) I will see how to fix that and then rebase.

danxuliu committed:
Fix use of data directory in integration tests

The data directory is not necessarily located at "../..". The proper
directory is now got by running "php console.php config:system:get
datadirectory".

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Make possible to provide "--tags=XXX" parameter to Behat

"--tags=XXX" limits the features or scenarios to be run to those
matching the tag filter expression.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Do not run download scenarios tagged as "large" in Drone

Large scenarios take too long to run, so they would be cancelled before
they were finished. Therefore, now they are not even run.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu commented:
As I suspected they take too long to run, so I have disabled those large integration tests in Drone.

MorrisJobke commented:
@danxuliu @rullzer Just to verify: using this I still can't extract the zip file on macOS 😕

danxuliu commented:
@MorrisJobke That is expected; a patch for PHPZipStreamer is needed, but it has not been added yet ;-)

MorrisJobke commented:
> @MorrisJobke That is expected; a patch for PHPZipStreamer is needed, but it has not been added yet ;-)

Problem now is, that his actually breaks the download of multiple files for macOS. Where are those fixes? Could we add them to this PR as well?

danxuliu commented:
@MorrisJobke 
>  Problem now is, that his actually breaks the download of multiple files for macOS.

Oops, this pull request should not have been in _To review_ state, sorry :-)

>  Where are those fixes? Could we add them to this PR as well?

The fixes are here: https://github.com/McNetic/PHPZipStreamer/pull/40 I do not know how @rullzer intended to add them to the server (in this pull request, on a different pull request and then rebase this one on top of it, or other approach).

rullzer commented:
Rebased on master to include #9101

MorrisJobke reviewed (approved):
Works on macOS 👍
